### PR TITLE
Feature/login-register

### PR DESCRIPTION
**Pull request by DennisRein**
- Add login and register view
- Add login and register controller
- Add authentication work flow
- Add loader animation
- Add popups
- Add validation of email format, unique user/email, etc.
- Update user_service for unique test
- Update main screen to check for cached client and navigate to right screen

As always I am thankful for any kind of critique.

### CODE REVIEW

File: fridgify/lib/utils/validator.dart [hunk 1]
```diff
@@ -0,0 +1,86 @@
+class Validator {
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:1]: I found a few mistakes in the validation messages and i think we should have consistant wording and capitalization. However a few things are debatable. I point out grammar mistakes i catched and give a few suggestions on how i would write theses messages

File: fridgify/lib/utils/validator.dart [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  static bool mailNotUnique = false;
+  static bool doNotMatch = true;
+
+  static String validateMail(String text) {
+    Pattern pattern =
+        r'^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$';
+    RegExp regex = new RegExp(pattern);
+
+    if(text.isEmpty) {
+      return 'Please enter an E-Mail';
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:12]: I would prefer email not to be hyphenated and written in lowercase.

File: fridgify/lib/utils/validator.dart [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    //RegExp exp = new RegExp(r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)");
+    //if(!exp.hasMatch(text))
+    //  return 'Please enter a valid E-Mail';
+    return null;
+  }
+
+  static String validateUser(String text) {
+    if(text.isEmpty) {
+      return 'Please enter an Username';
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:30]: Username should be lowercase in my opinion

File: fridgify/lib/utils/validator.dart [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    if(userNotUnique) {
+      return 'Username already used';
+    }
+    return null;
+  }
+
+  static String validateName(String text) {
+    if(text.isEmpty) {
+      return 'Please enter an Name';
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:40]: Should be "a name" and lowercase.

File: fridgify/lib/utils/validator.dart [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    if(userNotUnique) {
+      return 'Username already used';
+    }
+    return null;
+  }
+
+  static String validateFirst(String text) {
+    if(text.isEmpty) {
+      return 'Please enter an First Name';
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:50]: Should be "enter a first name"

File: fridgify/lib/utils/validator.dart [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    if(userNotUnique) {
+      return 'Username already used';
+    }
+    return null;
+  }
+
+  static String validateDate(String text) {
+    if(text.isEmpty) {
+      return 'Please enter an Birthdate';
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:60]: "a birthdate" in my opinion

File: fridgify/lib/utils/validator.dart [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    if(userNotUnique) {
+      return 'Username already used';
+    }
+    return null;
+  }
+
+  static String validatePassword(String text) {
+    if(text.isEmpty) {
+      return 'Please enter a Password!';
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:70]: Lowercase and i think the exclamation mark should either be removed or each message should have one. I think it should be removed

File: fridgify/lib/utils/validator.dart [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+      return 'Your password is too short';
+    return null;
+  }
+
+  static String repeatValidatePassword(String text) {
+    if(text.isEmpty) {
+      return 'Please repeat the Password!';
+    }
+    if(doNotMatch) {
+      return 'Password do not match!';
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:82]: Should be "Password doesn't match

File: fridgify/lib/utils/validator.dart [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+      return 'Please enter a Password!';
+    }
+    if(text.length < 6)
+      return 'Your password is too short';
+    return null;
+  }
+
+  static String repeatValidatePassword(String text) {
+    if(text.isEmpty) {
+      return 'Please repeat the Password!';
```
Comment (JoschuaGoetz) [inline fridgify/lib/utils/validator.dart:79]: Lowercase

File: fridgify/lib/utils/validator.dart (hunk 7 above)
Comment (DennisRein) [reply fridgify/lib/utils/validator.dart:70]: Thanks for mentioning these things. I'll correct them all. These mistakes happened when I was mindlessly copying and pasting them.

**Comment by DennisRein** [comment]
One last okay would be awesome. Thanks for the correction.

**Pull request merged**